source/common/upstream/ring_hash_lb.cc

```diff
@@ -58,36 +64,118 @@ HostConstSharedPtr RingHashLoadBalancer::Ring::chooseHost(uint64_t h) const {
+
 using HashFunction = envoy::api::v2::Cluster_RingHashLbConfig_HashFunction;
 RingHashLoadBalancer::Ring::Ring(
+    const HostSet& host_set, bool in_panic,
     const absl::optional<envoy::api::v2::Cluster::RingHashLbConfig>& config,
-    const HostVector& hosts) {
+    RingHashLoadBalancerStats& stats)
+    : stats_(stats) {
   ENVOY_LOG(trace, "ring hash: building ring");
-  if (hosts.empty()) {
-    return;
-  }
 
-  // Currently we specify the minimum size of the ring, and determine the replication factor
-  // based on the number of hosts. It's possible we might want to support more sophisticated
-  // configuration in the future.
-  // NOTE: Currently we keep a ring for healthy hosts and unhealthy hosts, and this is done per
-  //       thread. This is the simplest implementation, but it's expensive from a memory
-  //       standpoint and duplicates the regeneration computation. In the future we might want
-  //       to generate the rings centrally and then just RCU them out to each thread. This is
-  //       sufficient for getting started.
   const uint64_t min_ring_size =
-      config ? PROTOBUF_GET_WRAPPED_OR_DEFAULT(config.value(), minimum_ring_size, 1024) : 1024;
+      config
+          ? PROTOBUF_GET_WRAPPED_OR_DEFAULT(config.value(), minimum_ring_size, DefaultMinRingSize)
+          : DefaultMinRingSize;
+  const uint64_t max_ring_size =
+      config
+          ? PROTOBUF_GET_WRAPPED_OR_DEFAULT(config.value(), maximum_ring_size, DefaultMaxRingSize)
+          : DefaultMaxRingSize;
 
-  uint64_t hashes_per_host = 1;
-  if (hosts.size() < min_ring_size) {
-    hashes_per_host = min_ring_size / hosts.size();
-    if ((min_ring_size % hosts.size()) != 0) {
-      hashes_per_host++;
-    }
+  // Sanity-check ring size bounds.
+  if (min_ring_size > max_ring_size) {
+    throw EnvoyException(fmt::format("ring hash: minimum_ring_size ({}) > maximum_ring_size ({})",
+                                     min_ring_size, max_ring_size));
   }
 
-  ENVOY_LOG(info, "ring hash: min_ring_size={} hashes_per_host={}", min_ring_size, hashes_per_host);
-  ring_.reserve(hosts.size() * hashes_per_host);
+  // Normalize weights, such that the sum of all weights = 1.
+  std::unordered_map<HostConstSharedPtr, double> normalized_weights;
+  double min_normalized_weight = 1.0;
+  normalizeWeights(host_set, in_panic, normalized_weights, min_normalized_weight);
+
+  // We can't do anything sensible with no hosts.
+  if (normalized_weights.empty()) {
+    return;
+  }
+
+  // Scale up the number of hashes per host such that the least-weighted host gets a whole number
+  // of hashes on the ring. Other hosts might not end up with whole numbers, and that's fine (the
+  // ring-building algorithm below can handle this). This preserves the original implementation's
+  // behavior: when weights aren't provided, all hosts should get an equal number of hashes. In the
+  // case where this number exceeds the max_ring_size, it's scaled back down to fit.
```

Comment:
If we have a ring that has max/min size 1024 and we have 1 host in a locality with locality weight 127 and 1023 hosts in another locality with locality weight 1, what happens? It seems like we would have each host populated at least once on the ring, which means we have 1024 hosts, but the locality weighting is effectively ignored.

Reply:
Here's what happens in the case you describe:
* First, the single host in the LW=127 locality will have a normalized weight of 127/128, and each host in the LW=1 locality will have a normalized weight of 1/128 x 1/1023.
* Then we compute the "scale" (I'll assign dummy variable names to the parts of that expression):
    * a = ceil(min_normalized_weight * min_ring_size) = 1
    * b = a / min_normalized_weight = 130944
    * c = max_ring_size = 1024
    * scale = min(b, c) = 1024
* Finally we build the ring. The heavily-weighted host will get 1024*127/128 = 1016 hashes on the ring, and eight of the other very lightly-weighted hosts will get one hash. The others will get zero.

Reply:
SG, I assume there is test coverage for something like this?

Reply:
Yup.

Reply:
Actually, one more question. If the lightly weighted hosts come first when assigning to the ring, won't they each get  a dedicated entry before we get around to dealing with the 127/128 weighted host?

Reply:
Discussed IRL, the fractional target avoids this, but need to figure out how to make things deterministic. Resolved.